### DISCUSSION TITLE
Conditional class properties based on morph relation

### DISCUSSION THREAD

**Discussion by attac-t** [spatie/laravel-data#373] [Q&A]
Let's say we have a `Person` model which has many `ContactNumber` and each contact number can belong to a `Business` or a `Customer` which is a nullable `morphTo` relation. The following is the structure of the data classes.

```php
class PersonData extends Data
{
  public function __construct (
   public int $id,
   public string $firstName,
   public string $lastName
  )
}
```

```php
class ContactNumberData extends Data
{
  public function __construct (
   public int $id,
   public string $countryCode,
   public string $number,
   public BusinessData $business,
   public CustomerData $customer
  )
}
```

Within the `ContactNumberData` I'd like to have a separate key for `business` and `customer` and conditionally populate based on the owner `morphTo` relation.

Now from within the `PersonController`, we'll load `persons.contactNumbers.owner` and return `PersonData::collection(Person::all());`.

What would be the cleanest way to construct this nested structure using the `::collection` method?

**Comment by rubenvanassche** [comment]
Use a custom creation method

**Reply by attac-t** [reply]
Are you referring to overriding the static `from(...$payloads)` method? Cause we would need this custom creation method to be invoked automatically when called from a parent's `:: collection()` method such as `PersonData::collection(Person::all())`.

**Reply by rubenvanassche** [reply]
Take a look at this: https://spatie.be/docs/laravel-data/v3/as-a-data-transfer-object/creating-a-data-object#content-magical-creation
